agent/xds/server.go

```diff
@@ -0,0 +1,418 @@
+package xds
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"sync/atomic"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	authz "github.com/envoyproxy/go-control-plane/envoy/service/auth/v2alpha"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/gogo/googleapis/google/rpc"
+	"github.com/gogo/protobuf/proto"
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/cache"
+	"github.com/hashicorp/consul/agent/connect"
+	"github.com/hashicorp/consul/agent/proxycfg"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// ADSStream is a shorter way of referring to this thing...
+type ADSStream = discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer
+
+// Resource types in xDS v2. These are copied from
+// envoyproxy/go-control-plane/pkg/cache/resource.go since we don't need any of
+// the rest of that package.
+const (
+	typePrefix   = "type.googleapis.com/envoy.api.v2."
+	EndpointType = typePrefix + "ClusterLoadAssignment"
+	ClusterType  = typePrefix + "Cluster"
+	RouteType    = typePrefix + "RouteConfiguration"
+	ListenerType = typePrefix + "Listener"
+
+	LocalAppClusterName   = "local_app"
+	LocalAgentClusterName = "local_agent"
+)
+
+// ACLResolverFunc is a shim to resolve ACLs. Since ACL enforcement is so far
+// entirely agent-local and all uses private methods this allows a simple shim
+// to be written in the agent package to allow resolving without tightly
+// coupling this to the agent.
+type ACLResolverFunc func(id string) (acl.ACL, error)
+
+// ConnectAuthz is the interface the agent needs to expose to be able to re-use
+// the authorization logic between both APIs.
+type ConnectAuthz interface {
+	// ConnectAuthorize is implemented by Agent.ConnectAuthorize
+	ConnectAuthorize(token string, req *structs.ConnectAuthorizeRequest) (authz bool, reason string, m *cache.ResultMeta, err error)
+}
+
+// ConfigManager is the interface the config manager implements to make testing
+// the server without a full mocker manager, state and cache. In normal use this
+// is satisfied directly by a pointer to the agent's proxycfg.Manager.
+type ConfigManager interface {
```

Comment:
Can you also explain what a ConfigManager is? To the new reviewer, it isn't obvious what it does until reading more.

Reply:
Hmm I tried to, it's just a shim to allow us to test without needing a full stack of mocked cache, local state and proxy manager. I'll try to reword it if that wasn't clear. Maybe it should read:

> ConfigManager is the interface we require to consume proxy config updates. It's satisfied normally by the agent's proxycfg.Manager, but allows easier testing without several layers of mocked cache, local state and proxycfg.Manager.